File: src/Tests.CSharp/Text/InMemoryTextTest.cs
using System;
using System.Linq;
using Shouldly;
using Xunit;

namespace Demian.Tests.Text
{
    public class InMemoryTextTest
    {
        [Theory]
        [InlineData("")]
        [InlineData("Hello, world.")]
        [InlineData("3.14")]
        [InlineData("Ready to set the world on fire, hehehehehe.")]
        public void Ctor_ShouldInitializeCharactersFromProvidedString(string value)
        {
            var text = new InMemoryText(value);

            text.Characters.ShouldBe(value.Select(x => new Character(x)));
        }

        [Fact]
        public void Ctor_ShouldThrowArgumentNullException_IfStringIsNull() =>
            Assert.Throws<ArgumentNullException>(() => new InMemoryText(null));

        [Theory]
        [InlineData("", "Hello!", 0)]
        [InlineData("I was here already.", "Hello!", 0)]
        [InlineData("012345", "6", 1)]
        [InlineData("012345", "6", 2)]
        [InlineData("012345", "6", 3)]
        [InlineData("012345", "666", 4)]
        [InlineData("012345", "666", 5)]
        public void Write_ShouldInsertTextAtSpecifiedIndex(string initial, string toInsert, int at)
        {
            var text = new InMemoryText(initial);
            var textAsString = text.AsString();

            text.Write(toInsert, at);
            textAsString = textAsString.Insert(at, toInsert);

            text.AsString().ShouldBe(textAsString);
        }

        [Fact]
        public void Write_ShouldFail_IfValueIsNull() =>
            new InMemoryText("1").Write(null, 0).Fail.ShouldBe(true);

        [Fact]
        public void Write_ShouldFail_IfIndexIsLessThanZero() =>
            new InMemoryText("1").Write("", -1).Fail.ShouldBe(true);
        
        [Fact]
        public void Write_ShouldFail_IfIndexIsGreaterThanText() =>
            new InMemoryText("1234").Write("", 5).Fail.ShouldBe(true);

        [Theory]
        [InlineData("Hello", 2, 0)]
        [InlineData("Hello", 3, 0)]
        [InlineData("Hello", 5, 0)]
        [InlineData("Hello", 4, 1)]
        [InlineData("Hello", 3, 2)]
        [InlineData("Hello", 2, 3)]
        public void Remove_ShouldRemovePartOfTextAtSpecifiedIndex(string initial, int countToRemove, int at)
        {
            var text = new InMemoryText(initial);
            var textAsString = text.AsString();

            text.Remove(countToRemove, at);
            textAsString = textAsString.Remove(at, countToRemove);

            text.AsString().ShouldBe(textAsString);
        }
        
        [Fact]
        public void Remove_ShouldFail_IfLengthIsNegative() =>
            new InMemoryText("1").Remove(-1, 0).Fail.ShouldBe(true);
        
        [Fact]
        public void Remove_ShouldFail_IfLengthIsZero() =>
            new InMemoryText("1").Remove(0, 0).Fail.ShouldBe(true);
        
        [Fact]
        public void Remove_ShouldFail_IfLengthIsOverflowText() =>
            new InMemoryText("1").Remove(2, 0).Fail.ShouldBe(true);
        
        [Fact]
        public void Remove_ShouldFail_IfIndexIsLessThanZero() =>
            new InMemoryText("1").Remove(1, -1).Fail.ShouldBe(true);
        
        [Fact]
        public void Remove_ShouldFail_IfIndexIsGreaterThanText() =>
            new InMemoryText("1234").Remove(1, 5).Fail.ShouldBe(true);
    }
}

File: src/Demian.Tests.FSharp/Text/InMemoryTextTest.fs
module Demian.Tests.FSharp.Text.InMemoryTextTest

open System.Linq
open Demian
open FsUnit
open Pocket.Common
open System;
open Xunit;

let at x = x
let in' x = x

let asText value = value |> InMemoryText
let characters (text:InMemoryText) = text.Characters.ToArray()
let asCharacters (value:string) = value.Select(fun x -> new Character(x)).ToArray()
let throwIfFail (result:Result) = if (result.Fail) then raise (() |> Exception)

let write part at value =
    let text = value |> asText
    text.Write(part, at) |> ignore
    text.AsString()
    
let writing part at value =
    (fun () -> (asText value).Write(part, at) |> throwIfFail |> ignore)
    
let remove length at value =
    let text = value |> asText
    text.Remove(length, at) |> ignore
    text.AsString()
    
let removing length at value =
    (fun () -> (asText value).Remove(length, at) |> throwIfFail |> ignore)

[<Fact>]
let ``Text should fail to construct from null``() =
    (fun () -> null |> asText |> ignore) |> shouldFail
    
[<Theory>]
[<InlineData("")>]
[<InlineData("Hello, world.")>]
[<InlineData("3.14")>]
[<InlineData("Ready to set the world on fire, hehehehehe.")>]
let ``Text constructed from string should contain same characters`` (value) =
    value |> asText |> characters |> should equal (value |> asCharacters)

[<Fact>]
let ``When write "-" at 0 in "Hello" it should equal "-Hello"``() =
    write "-" <| at 0 <| in' "Hello" |> should equal "-Hello"
        
[<Fact>]
let ``When write "-" at 1 in "Hello" it should equal "H-ello"``() =
    write "-" <| at 1 <| in' "Hello" |> should equal "H-ello"
    
[<Fact>]
let ``When write "-" at 5 in "Hello" it should equal "Hello-"``() =
    write "-" <| at 5 <| in' "Hello" |> should equal "Hello-"
    
[<Fact>]
let ``When write "--" at 0 in "Hello" it should equal "--Hello"``() =
    write "--" <| at 0 <| in' "Hello" |> should equal "--Hello"
    
[<Fact>]
let ``When write "--" at 1 in "Hello" it should equal "H--ello"``() =
    write "--" <| at 1 <| in' "Hello" |> should equal "H--ello"
    
[<Fact>]
let ``When write "--" at 5 in "Hello" it should equal "Hello--"``() =
    write "--" <| at 5 <| in' "Hello" |> should equal "Hello--"
    
[<Fact>]
let ``Writing null should fail``() =
    writing null <| at 0 <| in' "Hello" |> shouldFail
    
[<Fact>]
let ``Writing empty string at -1 should fail``() =
    writing "" <| at -1 <| in' "Hello" |> shouldFail
    
[<Fact>]
let ``Writing empty string at 5 in "1234" should fail``() =
    writing "" <| at 5 <| in' "1234" |> shouldFail
    
[<Fact>]
let ``Writing empty string at 5 in "1234" should fail2``() =
    writing "" <| at 5 <| in' "1234" |> shouldFail
    
[<Fact>]
let ``When remove 1 character at 0 in "Hello" it should equal "ello"``() =
    remove 1 <| at 0 <| in' "Hello" |> should equal "ello"
    
[<Fact>]
let ``When remove 1 character at 1 in "Hello" it should equal "Hllo"``() =
    remove 1 <| at 1 <| in' "Hello" |> should equal "Hllo"
    
[<Fact>]
let ``When remove 1 character at 4 in "Hello" it should equal "Hell"``() =
    remove 1 <| at 4 <| in' "Hello" |> should equal "Hell"
    
[<Fact>]
let ``When remove 3 characters at 0 in "Hello" it should equal "lo"``() =
    remove 3 <| at 0 <| in' "Hello" |> should equal "lo"
    
[<Fact>]
let ``When remove 3 characters at 1 in "Hello" it should equal "Ho"``() =
    remove 3 <| at 1 <| in' "Hello" |> should equal "Ho"

[<Fact>]
let ``When remove 3 characters at 2 in "Hello" it should equal "He"``() =
    remove 3 <| at 2 <| in' "Hello" |> should equal "He"
    
[<Fact>]
let ``When remove 5 characters at 0 in "Hello" it should equal ""``() =
    remove 5 <| at 0 <| in' "Hello" |> should equal ""
    
[<Fact>]
let ``Removing -1 at 0 in "Hello" should fail``() =
    removing -1 <| at 0 <| in' "Hello" |> shouldFail
    
[<Fact>]
let ``Removing 0 at 0 in "Hello" should fail``() =
    removing 0 <| at 0 <| in' "Hello" |> shouldFail
    
[<Fact>]
let ``Removing 6 at 0 in "1" should fail``() =
    removing 6 <| at 0 <| in' "Hello" |> shouldFail
    
[<Fact>]
let ``Removing 1 at -1 in "Hello" should fail``() =
    removing 1 <| at -1 <| in' "Hello" |> shouldFail
    
[<Fact>]
let ``Removing 1 at 5 in "Hello" should fail``() =
    removing 1 <| at 5 <| in' "Hello" |> shouldFail